File: biometric-local-agent/src/workers/heartbeat-worker.ts
import { config } from "../config";
import { logger } from "../logger";
import { AttendanceStore } from "../storage/attendance-store";
import { ConfigWorker } from "./config-worker";
import { sendHeartbeat } from "../live-server/send-heartbeat";

export class HeartbeatWorker {
  private static isRunning = false;

  /**
   * Runs the heartbeat sync
   */
  static async run(): Promise<void> {
    if (this.isRunning) return;
    this.isRunning = true;

    logger.info("HeartbeatWorker: Sending heartbeat report...");

    try {
      // 1. Get database log metrics
      const maxRetries = config.maxRetryCount;
      const metrics = AttendanceStore.getStatusSummary(maxRetries);

      // 2. Get active device count
      const devices = ConfigWorker.getLocalDevices();
      const activeCount = devices.filter((d) => d.isActive).length;

      // 3. Send heartbeat to live server
      const result = await sendHeartbeat({
        status: "ONLINE",
        pendingLogs: metrics.pending,
        failedLogs: metrics.failed,
        activeDevices: activeCount
      });

      if (result.success) {
        logger.info("HeartbeatWorker: Heartbeat sent successfully.");
      } else {
        logger.warn(`HeartbeatWorker: Heartbeat failed: ${result.error}`);
      }
    } catch (err) {
      logger.error("HeartbeatWorker: Uncaught error in heartbeat task", err);
    } finally {
      this.isRunning = false;
    }
  }
}
export default HeartbeatWorker;

File: ts-biomatrics/app/workers/heartbeat_worker.py
from PySide6.QtCore import QThread, Signal
from app.database.database import SessionLocal
from app.database.models import RawBiometricLog, BiometricDevice
from app.services.queue_service import QueueService
from app.core.logger import get_logger

logger = get_logger("Heartbeat Worker")

class HeartbeatWorker(QThread):
    """
    Background worker to send a gateway heartbeat to the cloud.
    """
    finished = Signal(bool, str) # success, message

    def run(self):
        logger.info("HeartbeatWorker: Preparing gateway heartbeat payload...")
        db = SessionLocal()
        client = QueueService.get_api_client(db)
        
        if not client:
            msg = "Gateway not configured. Heartbeat skipped."
            logger.warning(msg)
            self.finished.emit(False, msg)
            db.close()
            return

        try:
            # Gather statistics for payload
            pending_logs = db.query(RawBiometricLog).filter(RawBiometricLog.sync_status == "PENDING").count()
            failed_logs = db.query(RawBiometricLog).filter(RawBiometricLog.sync_status == "FAILED").count()
            active_devices = db.query(BiometricDevice).filter(BiometricDevice.is_enabled == True).count()

            success, error_msg = client.send_heartbeat(
                status="ONLINE",
                pending_logs=pending_logs,
                failed_logs=failed_logs,
                active_devices=active_devices
            )

            if success:
                logger.info("HeartbeatWorker: Heartbeat successfully sent.")
                self.finished.emit(True, "Heartbeat Successful")
            else:
                logger.warning(f"HeartbeatWorker: Heartbeat failed: {error_msg}")
                # Save to retry queue
                QueueService.enqueue(
                    db=db,
                    operation_type="HEARTBEAT",
                    payload={
                        "status": "ONLINE",
                        "pendingLogs": pending_logs,
                        "failedLogs": failed_logs,
                        "activeDevices": active_devices
                    }
                )
                self.finished.emit(False, f"Heartbeat failed: {error_msg}")

        except Exception as e:
            logger.error(f"HeartbeatWorker: Error during execution: {e}")
            self.finished.emit(False, str(e))
        finally:
            db.close()
